torch/testing/_internal/common_utils.py

```diff
@@ -801,204 +801,304 @@ def is_integral(self, dtype):
+                                " different devices. Got device {0} and"
+                                " {1}.".format(a.device, b.device))
+
+        # Compares tensors of different devices on the CPU
+        if a.device != b.device:
+            a = a.cpu()
+            b = b.cpu()
+
+        # Checks size matches
+        if a.size() != b.size():
+            return False
+
+        # Checks dtype (if exact_dtype)
+        if exact_dtype and a.dtype is not b.dtype:
+            return False
+
+        # Acquires atol and rtol
+        # Note: atol can be overriden by a class's precision value. This
+        # legacy behavior allows generated tests to customize their precision.
+        rtol = rtol if rtol is not None else max(self.dtype_precisions.get(a.dtype, (0, 0))[0],
+                                                 self.dtype_precisions.get(b.dtype, (0, 0))[0])
+        atol = atol if atol is not None else max(self.dtype_precisions.get(a.dtype, (0, 0))[1],
+                                                 self.dtype_precisions.get(b.dtype, (0, 0))[1])
+        atol = max(atol, self.precision)
+
+
+        # Special-cases zero rtol and atol less than one since it allows
+        # comparing integer tensors using identity.
+        allow_ints = False
+        if (rtol == 0 and atol < 1):
+            allow_ints = True
+
+        dtype = self.compare_type(a, b, allow_ints=allow_ints)
+
+        # Converts to comparison dtype
+        a = a.to(dtype)
+        b = b.to(dtype)
+
+        # Integer (including bool) comparisons are identity comparisons
+        if self.is_integral(dtype):
+            return (a == b).all().item()
+
+        # Compares complex tensors' real and imaginary parts separately.
+        # Note: this is intentionally divergent from complex torch.isclose
+        # which implements a more mathematically natural notion of "closeness"
+        # for complex numbers.
+        if a.is_complex():
+            float_dtype = torch.float32 if a.dtype == torch.complex64 else torch.float64
+            a_real = a.copy_real().to(float_dtype)
+            b_real = b.copy_real().to(float_dtype)
+            real_result = self.compareTensors(a_real, b_real, rtol=rtol, atol=atol,
+                                              equal_nan=equal_nan, exact_dtype=exact_dtype,
+                                              exact_device=exact_device)
+
+            a_imag = a.copy_imag().to(float_dtype)
+            b_imag = b.copy_imag().to(float_dtype)
+            imag_result = self.compareTensors(a_imag, b_imag, rtol=rtol, atol=atol,
+                                              equal_nan=equal_nan, exact_dtype=exact_dtype,
+                                              exact_device=exact_device)
+
+            return real_result and imag_result
+
+        # All other types use torch.allclose directly
+        return torch.allclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan)
```

Comment:
by switching to is_close in the previous PR, when message=None we'd get the actual indices and values that mis-matched within a tensor. I'd like to preserve that behavior since it's often difficult to debug tests that just say "true != false."

Reply:
e.g. `torch.testing.assert_allclose(torch.tensor([1., 1.]), torch.tensor([1., 4.]))`
gives:
`AssertionError: Not within tolerance rtol=0.0001 atol=1e-05 at input[1] (1.0 vs. 4.0) and 0 other locations (50.00%)`

I think we had this behavior on master at least briefly. I'm not sure if I had it wrong or it regressed since i landed it, but not seeing it now.

Reply:
No worries. I can add it. I hope we kill torch.testing.allclose in the future, anyway, to keep assertEqual the canonical method of comparing tensors.

Reply:
It's possible users rely on allclose in their own unit tests, in libraries relying on PyTorch.  I believe numpy has an equivalent and my impression is we'd discourage users from creating instances of `torch.testing._internal.common_utils.TestCase`

Reply:
That's interesting. With the generic test framework we require people to subclass our TestCase. To support people not subclassing the TestCase we could encourage them to call compareTensors directly?